EasyPost/_base/EasyPostClient.cs

```diff
@@ -0,0 +1,150 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Reflection;
+using System.Threading.Tasks;
+using EasyPost.Http;
+using EasyPost.Models.API;
+using EasyPost.Utilities;
+using RestSharp;
+
+namespace EasyPost._base
+{
+    public abstract class EasyPostClient
+    {
+        public readonly ClientConfiguration Configuration;
+
+        private readonly RestClient _restClient;
+
+        // TODO: Revisit ApiVersion here vs in Request() when we decide how we want to handle accessing beta features
+        /// <summary>
+        ///     Constructor for the EasyPost client.
+        /// </summary>
+        /// <param name="apiKey">API key to use with this client.</param>
+        /// <param name="apiVersion">API version to use with this client.</param>
+        /// <param name="baseUrl">Base URL to use with this client. This will override `apiVersion`</param>
+        /// <param name="customHttpClient">
+        ///     Custom HttpClient to pass into RestSharp if needed. Mostly for debug purposes, not
+        ///     advised for general use.
+        /// </param>
+        protected EasyPostClient(string apiKey, ApiVersion? apiVersion = null, string? baseUrl = null, HttpClient? customHttpClient = null)
+        {
+            ServicePointManager.SecurityProtocol |= Security.GetProtocol();
+            Configuration = new ClientConfiguration(apiKey, apiVersion ?? ApiVersion.General, baseUrl, customHttpClient);
+
+            RestClientOptions clientOptions = new RestClientOptions
+            {
+                Timeout = Configuration.ConnectTimeoutMilliseconds,
+                BaseUrl = new Uri(Configuration.ApiBase),
+                UserAgent = Configuration.UserAgent
+            };
+
+            _restClient = customHttpClient != null ? new RestClient(customHttpClient, clientOptions) : new RestClient(clientOptions);
+        }
+
+        /// <summary>
+        ///     Execute a request against the EasyPost API.
+        /// </summary>
+        /// <typeparam name="T">Type of object to deserialize response data into.</typeparam>
+        /// <returns>An instance of a T type object.</returns>
+        internal async Task<T> Request<T>(Method method, string url, Dictionary<string, object>? parameters = null, string? rootElement = null, ApiVersion? apiVersion = null) where T : class
+        {
+            // Build the request
+            Request request = new Request(url, method, parameters, rootElement, apiVersion);
+            RestRequest restRequest = PrepareRequest(request);
+
+            // Execute the request
+            RestResponse<T> response = await _restClient.ExecuteAsync<T>(restRequest);
+
+            // Check the response's status code
+            if (!response.IsSuccessful)
+            {
+                // HTTP request threw an error (non-2xx response)
```

Comment:
3xx status codes can be successful, does the `IsSuccessful` account for those in the off chance we ever used them? We should really only realize a failure on `4xx` or `5xx`s.

Reply:
No, it only accounts for 2xx: https://docs.microsoft.com/en-us/uwp/api/windows.web.http.httpresponsemessage.issuccessstatuscode?view=winrt-22621#property-value

Reply:
Resolved.

Reply:
If so, the comment needs to be updated too

Reply:
Resolved.